File: src/backend/app/models/content.py
"""
Content models for the CMS system
"""
from enum import Enum
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Enum as SQLEnum, Boolean, UniqueConstraint, BigInteger
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from app.db.base import Base
from app.models.role import UserRole

class ContentType(str, Enum):
    """
    Enumeration of content types in the system
    """
    TEXT = "text"
    HTML = "html"
    MARKDOWN = "markdown"
    FILE = "file"

class ContentStatus(str, Enum):
    """
    Enumeration of content status values in the system
    """
    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"

class Content(Base):
    """
    Core content model representing all content items in the CMS
    """
    __tablename__ = "content"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(255), nullable=False, index=True)
    description = Column(String(1000), nullable=True)
    body = Column(Text, nullable=True)
    content_type = Column(SQLEnum(ContentType), nullable=False)
    status = Column(SQLEnum(ContentStatus), nullable=False, default=ContentStatus.DRAFT)
    
    # Ownership and audit
    created_by = Column(Integer, ForeignKey("user.id"), nullable=False)
    updated_by = Column(Integer, ForeignKey("user.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), nullable=False)
    
    # Relationships
    creator = relationship("User", foreign_keys=[created_by], backref="created_content")
    updater = relationship("User", foreign_keys=[updated_by], backref="updated_content")
    files = relationship("ContentFile", back_populates="content", cascade="all, delete-orphan")
    permissions = relationship("ContentPermission", back_populates="content", cascade="all, delete-orphan")

class ContentFile(Base):
    """
    Model for files attached to content items
    """
    __tablename__ = "content_file"

    id = Column(Integer, primary_key=True, index=True)
    content_id = Column(Integer, ForeignKey("content.id"), nullable=False)
    filename = Column(String(255), nullable=False)
    file_path = Column(String(1000), nullable=False)
    file_size = Column(BigInteger, nullable=False)
    mime_type = Column(String(127), nullable=False)
    uploaded_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    uploaded_by = Column(Integer, ForeignKey("user.id"), nullable=False)
    
    # Relationships
    content = relationship("Content", back_populates="files")
    uploader = relationship("User", backref="uploaded_files")

class ContentPermission(Base):
    """
    Model for role-based content permissions
    """
    __tablename__ = "content_permission"

    id = Column(Integer, primary_key=True, index=True)
    content_id = Column(Integer, ForeignKey("content.id"), nullable=False)
    role = Column(SQLEnum(UserRole), nullable=False)
    can_view = Column(Boolean, default=False, nullable=False)
    can_edit = Column(Boolean, default=False, nullable=False)
    can_delete = Column(Boolean, default=False, nullable=False)
    
    # Relationships
    content = relationship("Content", back_populates="permissions")
    
    # Composite unique constraint
    __table_args__ = (
        UniqueConstraint('content_id', 'role', name='uq_content_role'),
    )

File: src/frontend/src/types/content.ts
/**
 * Content-related type definitions for the CMS
 */

// Content status options
export enum ContentStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
  ARCHIVED = 'archived',
}

// Content type options
export enum ContentType {
  TEXT = 'text',
  HTML = 'html',
  MARKDOWN = 'markdown',
  FILE = 'file',
}

// Base content model
export interface Content {
  id: number;
  title: string;
  description: string | null;
  body: string | null;
  content_type: ContentType;
  status: ContentStatus;
  created_by: number;
  updated_by: number;
  created_at: string;
  updated_at: string;
}

// Content with relations (used for detailed views)
export interface ContentWithRelations extends Content {
  files?: ContentFile[];
  permissions?: ContentPermission[];
  creator?: {
    id: number;
    username: string;
  };
  updater?: {
    id: number;
    username: string;
  };
}

// Content creation payload
export interface ContentCreate {
  title: string;
  description?: string;
  body?: string;
  content_type: ContentType;
  status: ContentStatus;
}

// Content update payload
export interface ContentUpdate {
  title?: string;
  description?: string;
  body?: string;
  content_type?: ContentType;
  status?: ContentStatus;
}

// Content file model
export interface ContentFile {
  id: number;
  content_id: number;
  filename: string;
  file_path: string;
  file_size: number;
  mime_type: string;
  uploaded_at: string;
  uploaded_by: number;
}

// Permission level
export interface ContentPermission {
  id: number;
  content_id: number;
  role: string;
  can_view: boolean;
  can_edit: boolean;
  can_delete: boolean;
}

// Content query parameters
export interface ContentQueryParams {
  status?: ContentStatus;
  content_type?: ContentType;
  search?: string;
  page?: number;
  limit?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

// Content list response
export interface ContentListResponse {
  items: Content[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}
